File: implementations/go/client/internal/client/stun_message.go
package client

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha1"
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"net"
)

const (
	stunMagicCookie        = 0x2112A442
	stunHeaderBytes        = 20
	stunTransactionIDBytes = 12

	stunBindingRequest          = 0x0001
	stunBindingSuccess          = 0x0101
	stunBindingError            = 0x0111
	stunAllocateRequest         = 0x0003
	stunAllocateSuccess         = 0x0103
	stunAllocateError           = 0x0113
	stunRefreshRequest          = 0x0004
	stunRefreshSuccess          = 0x0104
	stunRefreshError            = 0x0114
	stunCreatePermissionRequest = 0x0008
	stunCreatePermissionSuccess = 0x0108
	stunCreatePermissionError   = 0x0118
	stunChannelBindRequest      = 0x0009
	stunChannelBindSuccess      = 0x0109
	stunChannelBindError        = 0x0119
	stunSendIndication          = 0x0016
	stunDataIndication          = 0x0017

	stunAttrMappedAddress      = 0x0001
	stunAttrChangeRequest      = 0x0003
	stunAttrErrorCode          = 0x0009
	stunAttrUnknownAttributes  = 0x000A
	stunAttrLifetime           = 0x000D
	stunAttrChannelNumber      = 0x000C
	stunAttrXorPeerAddress     = 0x0012
	stunAttrData               = 0x0013
	stunAttrXorRelayedAddress  = 0x0016
	stunAttrRequestedTransport = 0x0019
	stunAttrXorMappedAddress   = 0x0020
	stunAttrSoftware           = 0x8022
	stunAttrResponseOrigin     = 0x802B
	stunAttrOtherAddress       = 0x802C

	stunTransportUDP = 17

	stunAttrUsername         = 0x0006
	stunAttrMessageIntegrity = 0x0008
	stunAttrRealm            = 0x0014
	stunAttrNonce            = 0x0015
)

type stunAttribute struct {
	Type  uint16
	Value []byte
}

type stunMessage struct {
	Type          uint16
	TransactionID [stunTransactionIDBytes]byte
	Attributes    []stunAttribute
}

func newStunMessage(typ uint16, tx [stunTransactionIDBytes]byte, attrs ...stunAttribute) stunMessage {
	return stunMessage{Type: typ, TransactionID: tx, Attributes: attrs}
}

func newStunTransactionID() [stunTransactionIDBytes]byte {
	var tx [stunTransactionIDBytes]byte
	_, _ = rand.Read(tx[:])
	return tx
}

func stunTransactionHex(tx [stunTransactionIDBytes]byte) string {
	return hex.EncodeToString(tx[:])
}

func looksLikeStun(packet []byte) bool {
	if len(packet) < stunHeaderBytes || packet[0]&0xC0 != 0 {
		return false
	}
	length := int(binary.BigEndian.Uint16(packet[2:4]))
	cookie := binary.BigEndian.Uint32(packet[4:8])
	return cookie == stunMagicCookie && length+stunHeaderBytes == len(packet)
}

func parseStunMessage(packet []byte) (*stunMessage, error) {
	if !looksLikeStun(packet) {
		return nil, fmt.Errorf("not a stun message")
	}
	length := int(binary.BigEndian.Uint16(packet[2:4]))
	end := stunHeaderBytes + length
	message := &stunMessage{Type: binary.BigEndian.Uint16(packet[:2])}
	copy(message.TransactionID[:], packet[8:20])
	for offset := stunHeaderBytes; offset < end; {
		if end-offset < 4 {
			return nil, fmt.Errorf("truncated stun attribute")
		}
		attrType := binary.BigEndian.Uint16(packet[offset : offset+2])
		attrLen := int(binary.BigEndian.Uint16(packet[offset+2 : offset+4]))
		offset += 4
		if attrLen > end-offset {
			return nil, fmt.Errorf("invalid stun attribute length")
		}
		value := append([]byte(nil), packet[offset:offset+attrLen]...)
		message.Attributes = append(message.Attributes, stunAttribute{Type: attrType, Value: value})
		offset += attrLen + stunPadding(attrLen)
	}
	return message, nil
}

func (m stunMessage) bytes() []byte {
	return m.serialize(m.attributeBytes(), m.Attributes)
}

func (m stunMessage) bytesWithIntegrity(key []byte) []byte {
	if len(key) == 0 {
		return m.bytes()
	}
	beforeIntegrity := m.serialize(m.attributeBytes()+24, m.Attributes)
	mac := hmac.New(sha1.New, key)
	_, _ = mac.Write(beforeIntegrity)
	digest := mac.Sum(nil)
	result := make([]byte, len(beforeIntegrity)+24)
	copy(result, beforeIntegrity)
	offset := len(beforeIntegrity)
	binary.BigEndian.PutUint16(result[offset:offset+2], stunAttrMessageIntegrity)
	binary.BigEndian.PutUint16(result[offset+2:offset+4], uint16(len(digest)))
	copy(result[offset+4:], digest)
	return result
}

func (m stunMessage) attributeBytes() int {
	length := 0
	for _, attr := range m.Attributes {
		length += 4 + len(attr.Value) + stunPadding(len(attr.Value))
	}
	return length
}

func (m stunMessage) serialize(declaredLength int, attrs []stunAttribute) []byte {
	actualLength := 0
	for _, attr := range attrs {
		actualLength += 4 + len(attr.Value) + stunPadding(len(attr.Value))
	}
	packet := make([]byte, stunHeaderBytes+actualLength)
	binary.BigEndian.PutUint16(packet[:2], m.Type)
	binary.BigEndian.PutUint16(packet[2:4], uint16(declaredLength))
	binary.BigEndian.PutUint32(packet[4:8], stunMagicCookie)
	copy(packet[8:20], m.TransactionID[:])
	offset := stunHeaderBytes
	for _, attr := range attrs {
		binary.BigEndian.PutUint16(packet[offset:offset+2], attr.Type)
		binary.BigEndian.PutUint16(packet[offset+2:offset+4], uint16(len(attr.Value)))
		offset += 4
		copy(packet[offset:offset+len(attr.Value)], attr.Value)
		offset += len(attr.Value) + stunPadding(len(attr.Value))
	}
	return packet
}

func (m stunMessage) first(attrType uint16) (stunAttribute, bool) {
	for _, attr := range m.Attributes {
		if attr.Type == attrType {
			return attr, true
		}
	}
	return stunAttribute{}, false
}

func (m stunMessage) xorMappedAddress() (*net.UDPAddr, bool) {
	attr, ok := m.first(stunAttrXorMappedAddress)
	if !ok {
		return nil, false
	}
	return decodeStunXorAddress(attr.Value, m.TransactionID)
}

func (m stunMessage) mappedAddress() (*net.UDPAddr, bool) {
	attr, ok := m.first(stunAttrMappedAddress)
	if !ok {
		return nil, false
	}
	return decodeStunAddress(attr.Value)
}

func (m stunMessage) xorRelayedAddress() (*net.UDPAddr, bool) {
	attr, ok := m.first(stunAttrXorRelayedAddress)
	if !ok {
		return nil, false
	}
	return decodeStunXorAddress(attr.Value, m.TransactionID)
}

func (m stunMessage) xorPeerAddress() (*net.UDPAddr, bool) {
	attr, ok := m.first(stunAttrXorPeerAddress)
	if !ok {
		return nil, false
	}
	return decodeStunXorAddress(attr.Value, m.TransactionID)
}

func (m stunMessage) otherAddress() (*net.UDPAddr, bool) {
	attr, ok := m.first(stunAttrOtherAddress)
	if !ok {
		return nil, false
	}
	return decodeStunAddress(attr.Value)
}

func (m stunMessage) responseOrigin() (*net.UDPAddr, bool) {
	attr, ok := m.first(stunAttrResponseOrigin)
	if !ok {
		return nil, false
	}
	return decodeStunAddress(attr.Value)
}

func (m stunMessage) legacyXorOtherAddress() (*net.UDPAddr, bool) {
	attr, ok := m.first(stunAttrOtherAddress)
	if !ok {
		return nil, false
	}
	return decodeStunXorAddress(attr.Value, m.TransactionID)
}

func (m stunMessage) legacyXorResponseOrigin() (*net.UDPAddr, bool) {
	attr, ok := m.first(stunAttrResponseOrigin)
	if !ok {
		return nil, false
	}
	return decodeStunXorAddress(attr.Value, m.TransactionID)
}

func (m stunMessage) data() ([]byte, bool) {
	attr, ok := m.first(stunAttrData)
	if !ok {
		return nil, false
	}
	return append([]byte(nil), attr.Value...), true
}

func (m stunMessage) text(attrType uint16) string {
	attr, ok := m.first(attrType)
	if !ok {
		return ""
	}
	return string(attr.Value)
}

func (m stunMessage) errorCode() int {
	attr, ok := m.first(stunAttrErrorCode)
	if !ok || len(attr.Value) < 4 {
		return -1
	}
	return int(attr.Value[2]&0x07)*100 + int(attr.Value[3])
}

func (m stunMessage) unknownAttributes() []uint16 {
	attr, ok := m.first(stunAttrUnknownAttributes)
	if !ok {
		return nil
	}
	result := make([]uint16, 0, len(attr.Value)/2)
	for offset := 0; offset+2 <= len(attr.Value); offset += 2 {
		result = append(result, binary.BigEndian.Uint16(attr.Value[offset:offset+2]))
	}
	return result
}

func (m stunMessage) changeRequest() (changeIP, changePort, ok bool) {
	attr, found := m.first(stunAttrChangeRequest)
	if !found || len(attr.Value) != 4 {
		return false, false, false
	}
	flags := binary.BigEndian.Uint32(attr.Value)
	return flags&0x04 != 0, flags&0x02 != 0, true
}

func (m stunMessage) lifetimeSeconds(fallback int64) int64 {
	attr, ok := m.first(stunAttrLifetime)
	if !ok || len(attr.Value) != 4 {
		return fallback
	}
	return int64(binary.BigEndian.Uint32(attr.Value))
}

func (m stunMessage) channelNumber() (uint16, bool) {
	attr, ok := m.first(stunAttrChannelNumber)
	if !ok || len(attr.Value) != 4 {
		return 0, false
	}
	channel := binary.BigEndian.Uint16(attr.Value[:2])
	return channel, channel >= turnChannelMin && channel <= turnChannelMax
}

func newStunAttrXorPeerAddress(addr *net.UDPAddr, tx [stunTransactionIDBytes]byte) stunAttribute {
	return stunAttribute{Type: stunAttrXorPeerAddress, Value: encodeStunXorAddress(addr, tx)}
}

func stunAttrChannelNumberValue(channel uint16) stunAttribute {
	value := make([]byte, 4)
	binary.BigEndian.PutUint16(value[:2], channel)
	return stunAttribute{Type: stunAttrChannelNumber, Value: value}
}

func stunAttrDataValue(payload []byte) stunAttribute {
	return stunAttribute{Type: stunAttrData, Value: append([]byte(nil), payload...)}
}

func stunAttrLifetimeValue(seconds int64) stunAttribute {
	if seconds < 0 {
		seconds = 0
	}
	if seconds > 1<<32-1 {
		seconds = 1<<32 - 1
	}
	value := make([]byte, 4)
	binary.BigEndian.PutUint32(value, uint32(seconds))
	return stunAttribute{Type: stunAttrLifetime, Value: value}
}

func stunAttrRequestedUDPTransport() stunAttribute {
	return stunAttribute{Type: stunAttrRequestedTransport, Value: []byte{stunTransportUDP, 0, 0, 0}}
}

func stunAttrSoftwareValue(value string) stunAttribute {
	return stunAttribute{Type: stunAttrSoftware, Value: []byte(value)}
}

func stunAttrChangeRequestValue(changeIP, changePort bool) stunAttribute {
	flags := uint32(0)
	if changeIP {
		flags |= 0x04
	}
	if changePort {
		flags |= 0x02
	}
	value := make([]byte, 4)
	binary.BigEndian.PutUint32(value, flags)
	return stunAttribute{Type: stunAttrChangeRequest, Value: value}
}

func stunAttrResponseOriginValue(addr *net.UDPAddr) stunAttribute {
	return stunAttribute{Type: stunAttrResponseOrigin, Value: encodeStunAddress(addr)}
}

func stunAttrOtherAddressValue(addr *net.UDPAddr) stunAttribute {
	return stunAttribute{Type: stunAttrOtherAddress, Value: encodeStunAddress(addr)}
}

func stunAttrUnknownAttributesValue(types ...uint16) stunAttribute {
	value := make([]byte, len(types)*2)
	for index, attrType := range types {
		binary.BigEndian.PutUint16(value[index*2:index*2+2], attrType)
	}
	return stunAttribute{Type: stunAttrUnknownAttributes, Value: value}
}

func stunAttrUsernameValue(value string) stunAttribute {
	return stunAttribute{Type: stunAttrUsername, Value: []byte(value)}
}

func stunAttrRealmValue(value string) stunAttribute {
	return stunAttribute{Type: stunAttrRealm, Value: []byte(value)}
}

func stunAttrNonceValue(value string) stunAttribute {
	return stunAttribute{Type: stunAttrNonce, Value: []byte(value)}
}

func stunAttrErrorCodeValue(code int, reason string) stunAttribute {
	class := code / 100
	if class < 3 {
		class = 3
	} else if class > 6 {
		class = 6
	}
	number := code % 100
	if number < 0 {
		number = 0
	} else if number > 99 {
		number = 99
	}
	value := append([]byte{0, 0, byte(class), byte(number)}, []byte(reason)...)
	return stunAttribute{Type: stunAttrErrorCode, Value: value}
}

func encodeStunXorAddress(addr *net.UDPAddr, tx [stunTransactionIDBytes]byte) []byte {
	if addr == nil || addr.IP == nil {
		return nil
	}
	ip4 := addr.IP.To4()
	if ip4 != nil {
		value := make([]byte, 8)
		value[1] = 0x01
		binary.BigEndian.PutUint16(value[2:4], uint16(addr.Port^(stunMagicCookie>>16)))
		cookie := []byte{0x21, 0x12, 0xA4, 0x42}
		for i := 0; i < 4; i++ {
			value[4+i] = ip4[i] ^ cookie[i]
		}
		return value
	}
	ip16 := addr.IP.To16()
	if ip16 == nil {
		return nil
	}
	value := make([]byte, 20)
	value[1] = 0x02
	binary.BigEndian.PutUint16(value[2:4], uint16(addr.Port^(stunMagicCookie>>16)))
	mask := make([]byte, 16)
	binary.BigEndian.PutUint32(mask[:4], stunMagicCookie)
	copy(mask[4:], tx[:])
	for i := 0; i < 16; i++ {
		value[4+i] = ip16[i] ^ mask[i]
	}
	return value
}

func encodeStunAddress(addr *net.UDPAddr) []byte {
	if addr == nil || addr.IP == nil {
		return nil
	}
	ip4 := addr.IP.To4()
	if ip4 != nil {
		value := make([]byte, 8)
		value[1] = 0x01
		binary.BigEndian.PutUint16(value[2:4], uint16(addr.Port))
		copy(value[4:], ip4)
		return value
	}
	ip16 := addr.IP.To16()
	if ip16 == nil {
		return nil
	}
	value := make([]byte, 20)
	value[1] = 0x02
	binary.BigEndian.PutUint16(value[2:4], uint16(addr.Port))
	copy(value[4:], ip16)
	return value
}

func decodeStunAddress(value []byte) (*net.UDPAddr, bool) {
	if len(value) != 8 && len(value) != 20 {
		return nil, false
	}
	port := int(binary.BigEndian.Uint16(value[2:4]))
	switch value[1] {
	case 0x01:
		if len(value) != 8 {
			return nil, false
		}
		return &net.UDPAddr{IP: append(net.IP(nil), value[4:8]...), Port: port}, true
	case 0x02:
		if len(value) != 20 {
			return nil, false
		}
		return &net.UDPAddr{IP: append(net.IP(nil), value[4:20]...), Port: port}, true
	default:
		return nil, false
	}
}

func decodeStunXorAddress(value []byte, tx [stunTransactionIDBytes]byte) (*net.UDPAddr, bool) {
	if len(value) != 8 && len(value) != 20 {
		return nil, false
	}
	port := int(binary.BigEndian.Uint16(value[2:4]) ^ uint16(stunMagicCookie>>16))
	switch value[1] {
	case 0x01:
		if len(value) < 8 {
			return nil, false
		}
		cookie := []byte{0x21, 0x12, 0xA4, 0x42}
		ip := make(net.IP, 4)
		for i := 0; i < 4; i++ {
			ip[i] = value[4+i] ^ cookie[i]
		}
		return &net.UDPAddr{IP: ip, Port: port}, true
	case 0x02:
		if len(value) < 20 {
			return nil, false
		}
		mask := make([]byte, 16)
		binary.BigEndian.PutUint32(mask[:4], stunMagicCookie)
		copy(mask[4:], tx[:])
		ip := make(net.IP, 16)
		for i := 0; i < 16; i++ {
			ip[i] = value[4+i] ^ mask[i]
		}
		return &net.UDPAddr{IP: ip, Port: port}, true
	default:
		return nil, false
	}
}

func stunPadding(length int) int {
	return (4 - (length % 4)) % 4
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/StunMessage.cs
using System.Buffers.Binary;
using System.Net;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;

namespace Specus.Client.PeerMesh;

internal sealed class StunMessage
{
    public const uint MagicCookie = 0x2112A442;
    public const int HeaderBytes = 20;
    public const int TransactionIdBytes = 12;

    public const ushort BindingRequest = 0x0001;
    public const ushort BindingSuccess = 0x0101;
    public const ushort BindingError = 0x0111;
    public const ushort AllocateRequest = 0x0003;
    public const ushort AllocateSuccess = 0x0103;
    public const ushort AllocateError = 0x0113;
    public const ushort RefreshRequest = 0x0004;
    public const ushort RefreshSuccess = 0x0104;
    public const ushort RefreshError = 0x0114;
    public const ushort CreatePermissionRequest = 0x0008;
    public const ushort CreatePermissionSuccess = 0x0108;
    public const ushort CreatePermissionError = 0x0118;
    public const ushort ChannelBindRequest = 0x0009;
    public const ushort ChannelBindSuccess = 0x0109;
    public const ushort ChannelBindError = 0x0119;
    public const ushort SendIndication = 0x0016;
    public const ushort DataIndication = 0x0017;

    public const ushort AttrMappedAddress = 0x0001;
    public const ushort AttrChangeRequest = 0x0003;
    public const ushort AttrUsername = 0x0006;
    public const ushort AttrMessageIntegrity = 0x0008;
    public const ushort AttrErrorCode = 0x0009;
    public const ushort AttrUnknownAttributes = 0x000A;
    public const ushort AttrLifetime = 0x000D;
    public const ushort AttrChannelNumber = 0x000C;
    public const ushort AttrXorPeerAddress = 0x0012;
    public const ushort AttrData = 0x0013;
    public const ushort AttrRealm = 0x0014;
    public const ushort AttrNonce = 0x0015;
    public const ushort AttrXorRelayedAddress = 0x0016;
    public const ushort AttrRequestedTransport = 0x0019;
    public const ushort AttrXorMappedAddress = 0x0020;
    public const ushort AttrSoftware = 0x8022;
    public const ushort AttrResponseOrigin = 0x802B;
    public const ushort AttrOtherAddress = 0x802C;

    private const byte TransportUdp = 17;

    public StunMessage(ushort type, byte[] transactionId, IReadOnlyList<StunAttribute>? attributes = null)
    {
        Type = type;
        TransactionId = NormalizeTransactionId(transactionId);
        Attributes = attributes?.Select(item => new StunAttribute(item.Type, item.Value)).ToList() ?? [];
    }

    public ushort Type { get; }
    public byte[] TransactionId { get; }
    public IReadOnlyList<StunAttribute> Attributes { get; }
    public string TransactionIdHex => Convert.ToHexString(TransactionId).ToLowerInvariant();

    public static StunMessage Of(ushort type, byte[] transactionId, params StunAttribute[] attributes) =>
        new(type, transactionId, attributes);

    public static byte[] NewTransactionId()
    {
        var bytes = new byte[TransactionIdBytes];
        RandomNumberGenerator.Fill(bytes);
        return bytes;
    }

    public static bool LooksLike(ReadOnlySpan<byte> packet)
    {
        if (packet.Length < HeaderBytes || (packet[0] & 0xC0) != 0)
        {
            return false;
        }
        var declaredLength = BinaryPrimitives.ReadUInt16BigEndian(packet[2..4]);
        var cookie = BinaryPrimitives.ReadUInt32BigEndian(packet[4..8]);
        return cookie == MagicCookie && HeaderBytes + declaredLength == packet.Length;
    }

    public static StunMessage? Parse(ReadOnlySpan<byte> packet)
    {
        if (!LooksLike(packet))
        {
            return null;
        }
        var type = BinaryPrimitives.ReadUInt16BigEndian(packet[..2]);
        var length = BinaryPrimitives.ReadUInt16BigEndian(packet[2..4]);
        var end = HeaderBytes + length;
        var transactionId = packet[8..20].ToArray();
        var attributes = new List<StunAttribute>();
        var offset = HeaderBytes;
        while (offset < end)
        {
            if (end - offset < 4)
            {
                return null;
            }
            var attrType = BinaryPrimitives.ReadUInt16BigEndian(packet[offset..(offset + 2)]);
            var attrLength = BinaryPrimitives.ReadUInt16BigEndian(packet[(offset + 2)..(offset + 4)]);
            offset += 4;
            if (attrLength > end - offset)
            {
                return null;
            }
            attributes.Add(new StunAttribute(attrType, packet.Slice(offset, attrLength).ToArray()));
            offset += attrLength + Padding(attrLength);
        }
        return new StunMessage(type, transactionId, attributes);
    }

    public byte[] ToBytes()
    {
        return ToBytes(null);
    }

    public byte[] ToBytes(byte[]? messageIntegrityKey)
    {
        var length = Attributes.Sum(attr => 4 + attr.Value.Length + Padding(attr.Value.Length));
        if (messageIntegrityKey is { Length: > 0 })
        {
            var beforeIntegrity = Serialize(length + 24, Attributes);
            using var hmac = new HMACSHA1(messageIntegrityKey);
            var digest = hmac.ComputeHash(beforeIntegrity);
            var packetWithIntegrity = new byte[beforeIntegrity.Length + 24];
            beforeIntegrity.CopyTo(packetWithIntegrity, 0);
            var offset = beforeIntegrity.Length;
            BinaryPrimitives.WriteUInt16BigEndian(packetWithIntegrity.AsSpan(offset, 2), AttrMessageIntegrity);
            BinaryPrimitives.WriteUInt16BigEndian(packetWithIntegrity.AsSpan(offset + 2, 2), (ushort)digest.Length);
            digest.CopyTo(packetWithIntegrity.AsSpan(offset + 4));
            return packetWithIntegrity;
        }
        return Serialize(length, Attributes);
    }

    private byte[] Serialize(int declaredLength, IReadOnlyList<StunAttribute> attributes)
    {
        var length = attributes.Sum(attr => 4 + attr.Value.Length + Padding(attr.Value.Length));
        var packet = new byte[HeaderBytes + length];
        BinaryPrimitives.WriteUInt16BigEndian(packet.AsSpan(0, 2), Type);
        BinaryPrimitives.WriteUInt16BigEndian(packet.AsSpan(2, 2), (ushort)declaredLength);
        BinaryPrimitives.WriteUInt32BigEndian(packet.AsSpan(4, 4), MagicCookie);
        TransactionId.CopyTo(packet.AsSpan(8, TransactionIdBytes));
        var offset = HeaderBytes;
        foreach (var attr in attributes)
        {
            BinaryPrimitives.WriteUInt16BigEndian(packet.AsSpan(offset, 2), attr.Type);
            BinaryPrimitives.WriteUInt16BigEndian(packet.AsSpan(offset + 2, 2), (ushort)attr.Value.Length);
            offset += 4;
            attr.Value.CopyTo(packet.AsSpan(offset, attr.Value.Length));
            offset += attr.Value.Length + Padding(attr.Value.Length);
        }
        return packet;
    }

    public StunAttribute? First(ushort type) => Attributes.FirstOrDefault(attr => attr.Type == type);
    public string? TextAttribute(ushort type)
    {
        var value = First(type)?.Value;
        return value is { Length: > 0 } ? Encoding.UTF8.GetString(value) : null;
    }

    public int ErrorCode()
    {
        var value = First(AttrErrorCode)?.Value;
        if (value is not { Length: >= 4 })
        {
            return -1;
        }
        return (value[2] & 0x07) * 100 + value[3];
    }

    public IReadOnlyList<ushort> UnknownAttributes()
    {
        var value = First(AttrUnknownAttributes)?.Value;
        if (value is not { Length: >= 2 })
        {
            return [];
        }
        var result = new List<ushort>(value.Length / 2);
        for (var offset = 0; offset + 2 <= value.Length; offset += 2)
        {
            result.Add(BinaryPrimitives.ReadUInt16BigEndian(value.AsSpan(offset, 2)));
        }
        return result;
    }

    public StunChangeRequest? ChangeRequest()
    {
        var value = First(AttrChangeRequest)?.Value;
        if (value is not { Length: 4 })
        {
            return null;
        }
        var flags = BinaryPrimitives.ReadUInt32BigEndian(value);
        return new StunChangeRequest((flags & 0x04) != 0, (flags & 0x02) != 0);
    }

    public IPEndPoint? MappedAddress() => DecodeAddress(First(AttrMappedAddress)?.Value);
    public IPEndPoint? XorMappedAddress() => DecodeXorAddress(First(AttrXorMappedAddress)?.Value);
    public IPEndPoint? XorRelayedAddress() => DecodeXorAddress(First(AttrXorRelayedAddress)?.Value);
    public IPEndPoint? XorPeerAddress() => DecodeXorAddress(First(AttrXorPeerAddress)?.Value);
    public IPEndPoint? ResponseOrigin() => DecodeAddress(First(AttrResponseOrigin)?.Value);
    public IPEndPoint? OtherAddress() => DecodeAddress(First(AttrOtherAddress)?.Value);
    public IPEndPoint? LegacyXorResponseOrigin() => DecodeXorAddress(First(AttrResponseOrigin)?.Value);
    public IPEndPoint? LegacyXorOtherAddress() => DecodeXorAddress(First(AttrOtherAddress)?.Value);
    public byte[]? Data() => First(AttrData)?.Value.ToArray();

    public long LifetimeSeconds(long fallback)
    {
        var value = First(AttrLifetime)?.Value;
        return value is { Length: 4 } ? BinaryPrimitives.ReadUInt32BigEndian(value) : fallback;
    }

    public ushort? ChannelNumber()
    {
        var value = First(AttrChannelNumber)?.Value;
        if (value is not { Length: 4 })
        {
            return null;
        }
        var channel = BinaryPrimitives.ReadUInt16BigEndian(value);
        return channel is >= TurnChannelData.MinChannel and <= TurnChannelData.MaxChannel ? channel : null;
    }

    public bool RequestedUdpTransport()
    {
        var value = First(AttrRequestedTransport)?.Value;
        return value is { Length: > 0 } && value[0] == TransportUdp;
    }

    public static StunAttribute XorMappedAddress(IPEndPoint endpoint, byte[] transactionId) =>
        new(AttrXorMappedAddress, EncodeXorAddress(endpoint, transactionId));

    public static StunAttribute XorRelayedAddress(IPEndPoint endpoint, byte[] transactionId) =>
        new(AttrXorRelayedAddress, EncodeXorAddress(endpoint, transactionId));

    public static StunAttribute XorPeerAddress(IPEndPoint endpoint, byte[] transactionId) =>
        new(AttrXorPeerAddress, EncodeXorAddress(endpoint, transactionId));

    public static StunAttribute MappedAddress(IPEndPoint endpoint) =>
        new(AttrMappedAddress, EncodeAddress(endpoint));

    public static StunAttribute OtherAddress(IPEndPoint endpoint, byte[] transactionId) =>
        new(AttrOtherAddress, EncodeAddress(endpoint));

    public static StunAttribute ResponseOrigin(IPEndPoint endpoint, byte[] transactionId) =>
        new(AttrResponseOrigin, EncodeAddress(endpoint));

    public static StunAttribute ChangeRequest(bool changeIp, bool changePort)
    {
        var flags = (changeIp ? 0x04u : 0u) | (changePort ? 0x02u : 0u);
        var value = new byte[4];
        BinaryPrimitives.WriteUInt32BigEndian(value, flags);
        return new StunAttribute(AttrChangeRequest, value);
    }

    public static StunAttribute UnknownAttributes(params ushort[] types)
    {
        var normalized = types ?? [];
        var value = new byte[normalized.Length * 2];
        for (var index = 0; index < normalized.Length; index++)
        {
            BinaryPrimitives.WriteUInt16BigEndian(value.AsSpan(index * 2, 2), normalized[index]);
        }
        return new StunAttribute(AttrUnknownAttributes, value);
    }

    public static StunAttribute Data(byte[] payload) => new(AttrData, payload);

    public static StunAttribute ChannelNumber(ushort channel)
    {
        var value = new byte[4];
        BinaryPrimitives.WriteUInt16BigEndian(value, channel);
        return new StunAttribute(AttrChannelNumber, value);
    }

    public static StunAttribute Lifetime(long seconds)
    {
        var normalized = Math.Clamp(seconds, 0L, uint.MaxValue);
        var value = new byte[4];
        BinaryPrimitives.WriteUInt32BigEndian(value, (uint)normalized);
        return new StunAttribute(AttrLifetime, value);
    }

    public static StunAttribute RequestedUdpTransportAttribute() =>
        new(AttrRequestedTransport, [TransportUdp, 0, 0, 0]);

    public static StunAttribute Software(string value) =>
        new(AttrSoftware, Encoding.UTF8.GetBytes(value));

    public static StunAttribute Username(string value) =>
        new(AttrUsername, Encoding.UTF8.GetBytes(value));

    public static StunAttribute Realm(string value) =>
        new(AttrRealm, Encoding.UTF8.GetBytes(value));

    public static StunAttribute Nonce(string value) =>
        new(AttrNonce, Encoding.UTF8.GetBytes(value));

    public static StunAttribute ErrorCode(int code, string reason)
    {
        var klass = Math.Clamp(code / 100, 3, 6);
        var number = Math.Clamp(code % 100, 0, 99);
        var reasonBytes = Encoding.UTF8.GetBytes(reason);
        var value = new byte[4 + reasonBytes.Length];
        value[2] = (byte)klass;
        value[3] = (byte)number;
        reasonBytes.CopyTo(value.AsSpan(4));
        return new StunAttribute(AttrErrorCode, value);
    }

    private static IPEndPoint? DecodeAddress(byte[]? value)
    {
        if (value is null || (value.Length != 8 && value.Length != 20))
        {
            return null;
        }
        var port = BinaryPrimitives.ReadUInt16BigEndian(value.AsSpan(2, 2));
        if (value[1] == 0x01 && value.Length == 8)
        {
            return new IPEndPoint(new IPAddress(value.AsSpan(4, 4)), port);
        }
        if (value[1] == 0x02 && value.Length == 20)
        {
            return new IPEndPoint(new IPAddress(value.AsSpan(4, 16)), port);
        }
        return null;
    }

    private IPEndPoint? DecodeXorAddress(byte[]? value)
    {
        if (value is null || (value.Length != 8 && value.Length != 20))
        {
            return null;
        }
        var port = BinaryPrimitives.ReadUInt16BigEndian(value.AsSpan(2, 2)) ^ (MagicCookie >> 16);
        if (value[1] == 0x01 && value.Length >= 8)
        {
            var raw = new byte[4];
            Span<byte> cookie = [0x21, 0x12, 0xA4, 0x42];
            for (var i = 0; i < raw.Length; i++)
            {
                raw[i] = (byte)(value[4 + i] ^ cookie[i]);
            }
            return new IPEndPoint(new IPAddress(raw), (int)port);
        }
        if (value[1] == 0x02 && value.Length >= 20)
        {
            var mask = new byte[16];
            BinaryPrimitives.WriteUInt32BigEndian(mask.AsSpan(0, 4), MagicCookie);
            TransactionId.CopyTo(mask.AsSpan(4));
            var raw = new byte[16];
            for (var i = 0; i < raw.Length; i++)
            {
                raw[i] = (byte)(value[4 + i] ^ mask[i]);
            }
            return new IPEndPoint(new IPAddress(raw), (int)port);
        }
        return null;
    }

    private static byte[] EncodeXorAddress(IPEndPoint endpoint, byte[] transactionId)
    {
        var address = endpoint.Address.AddressFamily == AddressFamily.InterNetworkV6
            ? endpoint.Address.GetAddressBytes()
            : endpoint.Address.MapToIPv4().GetAddressBytes();
        var result = new byte[address.Length == 4 ? 8 : 20];
        result[1] = address.Length == 4 ? (byte)0x01 : (byte)0x02;
        BinaryPrimitives.WriteUInt16BigEndian(result.AsSpan(2, 2),
            (ushort)(endpoint.Port ^ (MagicCookie >> 16)));
        if (address.Length == 4)
        {
            Span<byte> cookie = [0x21, 0x12, 0xA4, 0x42];
            for (var i = 0; i < address.Length; i++)
            {
                result[4 + i] = (byte)(address[i] ^ cookie[i]);
            }
        }
        else
        {
            var mask = new byte[16];
            BinaryPrimitives.WriteUInt32BigEndian(mask.AsSpan(0, 4), MagicCookie);
            NormalizeTransactionId(transactionId).CopyTo(mask.AsSpan(4));
            for (var i = 0; i < address.Length; i++)
            {
                result[4 + i] = (byte)(address[i] ^ mask[i]);
            }
        }
        return result;
    }

    private static byte[] EncodeAddress(IPEndPoint endpoint)
    {
        var address = endpoint.Address.AddressFamily == AddressFamily.InterNetworkV6
            ? endpoint.Address.GetAddressBytes()
            : endpoint.Address.MapToIPv4().GetAddressBytes();
        var result = new byte[address.Length == 4 ? 8 : 20];
        result[1] = address.Length == 4 ? (byte)0x01 : (byte)0x02;
        BinaryPrimitives.WriteUInt16BigEndian(result.AsSpan(2, 2), (ushort)endpoint.Port);
        address.CopyTo(result.AsSpan(4));
        return result;
    }

    private static byte[] NormalizeTransactionId(byte[]? transactionId)
    {
        if (transactionId is { Length: TransactionIdBytes })
        {
            return transactionId.ToArray();
        }
        return NewTransactionId();
    }

    private static int Padding(int length) => (4 - length % 4) % 4;
}

internal sealed record StunAttribute(ushort Type, byte[] Value)
{
    public byte[] Value { get; init; } = Value.ToArray();
}

internal sealed record StunChangeRequest(bool ChangeIp, bool ChangePort);
